src/primitives/line/styled.rs

```diff
@@ -52,6 +59,104 @@ impl<C: PixelColor> StyledPixels<PrimitiveStyle<C>> for Line {
+    }
+    Ok(())
+}
+
+/// Compute the translation needed so that the dotted line fits the line as well as possible.
+///
+/// Rasterization of lines and dots causes positioning errors in the following situations :
+/// - when the dot size is even, the rasterized dot is shifted to the bottom right by 0.5px;
+/// - when the line has an even number of bresenham lines, the line is shifted by 0.5px
+///   to the left.
+///
+/// When both sources of error add up, the naive dot positioning can be improved
+/// (by translating the dots to the top left).
+fn start_dot_offset(line: &Line, dot_size: i32) -> Point {
+    if dot_size % 2 == 1 {
+        // The dot size is odd so the positioning is already ideal.
+        return line.start;
+    }
+
+    // Depending on the line orientation, the number of bresenham lines might be odd or even.
+    // The code below supposes that nearly horizontal/vertical lines are thicker on the left and
+    // that diagonals are symmetric. This is an approximation (there can be diagonal lines with
+    // an even number of bresenham lines).
+    //
+    // The code also makes the choice that the geometric center of the first dot should lie
+    // on the edge of the rasterized line.
+    //
+    // There's a drawing of the resulting alignment in test `start_dot_offset_matches_drawing`.
+    let mut start = line.start;
+    let delta = if start != line.end {
+        line.delta()
+    } else {
+        HORIZONTAL_LINE.delta()
+    };
+    if is_nearly_horizontal_or_vertical(line) {
+        let to_bottom_right = delta.dot_product(Point::new(1, 1));
+        let to_top_right = delta.dot_product(Point::new(1, -1));
+
+        if to_bottom_right > 0 {
+            start.y -= 1;
+        };
+        if to_top_right > 0 {
+            start.x -= 1;
+        };
+        start
+    } else {
+        let to_right = delta.dot_product(Point::new(1, 0));
+        let to_bottom = delta.dot_product(Point::new(0, 1));
+
+        if to_right > 0 {
+            start.x -= 1;
+        };
+        if to_bottom > 0 {
+            start.y -= 1;
+        };
+        start
+    }
+}
+
+fn is_nearly_horizontal_or_vertical(line: &Line) -> bool {
+    let mut points = Points::new(line);
+
+    let Some(second) = points.nth(1) else {
+        // The orientation of the degenerated line is horizontal.
+        return true;
+    };
+
+    let diff = second - line.start;
+    let sum = diff.abs().dot_product(Point::new(1, 1));
+
+    // If the first two pixels share an edge, the line is nearly horizontal or vertical.
+    sum == 1
+}
+
+fn extend_line_by_one_unit(line: &Line) -> Line {
+    let mut points = Points::new(line);
+
+    if let Some(second) = points.nth(1) {
+        Line::new(line.start, line.end + second - line.start)
+    } else {
+        // If there is only one point in the iterator, the line is expected to be reduced to a point.
+        // In that case we don't extend the line.
+        *line
+    }
```

Comment:
It took me a moment to realize that the if statement is the other way around than in the function above. I would suggest to use the same structure for both functions:

```suggestion
    let Some(second) = points.nth(1) else {
        // If there is only one point in the iterator, the line is expected to be reduced to a point.
        // In that case we don't extend the line.
        return *line;
    };
    
    Line::new(line.start, line.end + second - line.start)
```